src/vm_cpu.rs

```diff
@@ -316,35 +320,71 @@ impl VmPmuState {
     }
 
     /// Restores PMU counters from the saved state, reconfiguring and starting them if necessary.
+    /// Returns a computed value that should be be programmed into hcounteren
     /// This should be called in anticipation of an outbound context switch.
-    pub fn restore_counters(&mut self) {
+    pub fn restore_counters(&mut self) -> u64 {
+        fn get_hcounteren_bit(counter_index: u64) -> u64 {
+            let pmu_info = PmuInfo::get();
+            if let Some(info) = pmu_info.get_counter_info(counter_index) {
+                (1 << info.get_csr().checked_sub(CSR_CYCLE.into()).unwrap_or(32)) & 0xffff_ffff
```

Comment:
The `unwrap_or()` and the mask should be no-ops, and if they aren't they're hiding a bug.